src/encrypt.rs

```diff
@@ -0,0 +1,450 @@
+///
+///  The CDDL fragment that represents the above text for COSE_Encrypt0
+///  follows.
+///
+///  COSE_Encrypt0 = [
+///      Headers,
+///      ciphertext : bstr / nil,
+///  ]
+#[derive(Debug, Clone, Deserialize)]
+pub struct COSEEncrypt0 {
+    /// protected: empty_or_serialized_map,
+    protected: ByteBuf,
+    /// unprotected: HeaderMap
+    unprotected: HeaderMap,
+    /// ciphertext: bstr
+    /// The spec allows ciphertext to be nil and transported separately, but it's not useful at the
+    /// moment, so this is just a ByteBuf for simplicity.
+    ciphertext: ByteBuf,
+}
+
+impl Serialize for COSEEncrypt0 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(3))?;
+        seq.serialize_element(&self.protected)?;
+        seq.serialize_element(&self.unprotected)?;
+        seq.serialize_element(&self.ciphertext)?;
+        seq.end()
+    }
+}
+
+impl COSEEncrypt0 {
+    /// Creates a new instance of the COSE_Encrypt0 structure and encrypts the provided payload.
+    /// https://datatracker.ietf.org/doc/html/rfc8152#section-5.3
+    pub fn new(
+        payload: &[u8],
+        cipher: Cipher,
+        key: &[u8],
+    ) -> Result<Self, COSEError> {
+        let mut iv = vec![0; cipher.iv_len().unwrap()];
+        rand_bytes(&mut iv).unwrap();
+
+        let cose_alg = match COSEAlgorithm::from_nid(cipher.nid()) {
+            None => return Err(COSEError::UnsupportedError(
+                "Unsupported encryption algorithm"
+                    .to_string(),
+            )),
+            Some(v) => v
+        };
+        let cose_alg_value = cose_alg.value();
+        let mut protected = HeaderMap::new();
+        protected.insert(1.into(), CborValue::Integer(cose_alg_value as i128));
+        let mut unprotected = HeaderMap::new();
+        unprotected.insert(5.into(), CborValue::Bytes(iv.to_owned()));
+
+        let protected_bytes =
+            map_to_empty_or_serialized(&protected).map_err(COSEError::SerializationError)?;
+
+        let enc_structure = EncStructure::new_encrypt0(&protected_bytes)
+            .map_err(COSEError::SerializationError)?;
+
+        let mut tag = vec![0; cose_alg.tag_size()];
+        let mut ciphertext = encrypt_aead(cipher,
+            key,
+            Some(&iv[..]),
+            &enc_structure
+                .as_bytes()
+                .map_err(COSEError::SerializationError)?,
+            payload,
+            &mut tag).map_err(COSEError::EncryptionError)?;
+
+        ciphertext.append(&mut tag);
+
+        Ok(COSEEncrypt0{
+            protected: ByteBuf::from(protected_bytes),
+            unprotected,
+            ciphertext: ByteBuf::from(ciphertext)
+        })
+    }
+
+    /// Decrypt the ciphertext in the COSE_Encrypt0 structure.
+    /// https://datatracker.ietf.org/doc/html/rfc8152#section-5.3
+    pub fn decrypt(
```

Comment:
I think we might want to return the `protected` and `unprotected` buffers as well. If not from this one, maybe make an underlying function that returns it all and make this function only return the payload or something?

Reply:
changed to

```diff
     pub fn decrypt(
         &self,
         key: &[u8],
-    ) -> Result<Vec<u8>, COSEError> {
+    ) -> Result<(HeaderMap, &HeaderMap, Vec<u8>), COSEError> {
```